### DISCUSSION TITLE
Issues with 3D volume rendering of simulated cloud field

### DISCUSSION THREAD

**Discussion by mckenna-stanford-columbia** [pyvista/pyvista#3903] [Q&A] [answered]
### I am having issues getting a 3D volume to render properly. My guess is that this is entirely user-error. I am trying to render total water content to represent a simulated cloud field.  Following is all my code. The working grid is longitude x latitude x height (which I convert to meters for all 3 dimensions below). The 'twc' scalar field can probably just be represented as a random scalar field for reproducibility. The issue is that when I try to render using 'add_volume' using the grid and the scalar field (first rendered image below), it does not render the volume but rather something associated with the grid. This then changes the scalar field (twc) to the np.uint8 0-255 color range (something related to #475, I guess, though I think this was fixed?). If I then use 'add_volume' with just copies of the 'twc' scalar field, the scalar is replaced by the np.uint8 0-255 color range field, and although it produces what I'm looking for, it's not on the original grid and I'm guessing there's a more elegant way for me to do this.

#----------------------------------------------------------
# Imports
`import pyvista as pv`
`import numpy as np`
`import glob`
`import xarray`

# Print Report
`pv.Report()`
<img width="816" alt="Screen Shot 2023-01-30 at 11 10 36 AM" src="https://user-images.githubusercontent.com/85080500/215530920-45b2bf6e-46c3-41e7-8efd-23af4566753d.png">

# List files
`files = glob.glob('./twc*.nc')`
`file = files[0]`

# Read in file
`ncfile = xarray.open_dataset(file)`
`twc = np.array(ncfile['twc'].copy()) # total water content in g/m^3`
`z = np.array(ncfile['height'].copy()) # height in m AGL`
`lon = np.array(ncfile['lon'].copy()) # longitude `
`lat = np.array(ncfile['lat'].copy()) # latitude`
`ncfile.close()`

# Zero out small negative values resulting from interpolation
`twc[twc < 0.] = 0.0`
`twc[np.isnan(twc)] = 0.0`
`lon_1d = lon[:,0]`
`lat_1d = lat[0,:]`

# Variable looks more coherent in logarithmic scaling
`dumid = np.where(twc > 0.)`
`twc[twc == 0.] = np.nan`
`twc[dumid] = np.log10(twc[dumid])`

# Look at scalar limits (all looks good)
`print(np.nanmax(twc))`
`print(np.nanmin(twc))`

1.0488778143926158
-21.14496502505222

# Create grid

`x = np.arange(0,len(lon_1d),1)*600 # 600-m grid spacing`
`y = np.arange(0,len(lat_1d),1)*600 # 600-m grid spacing`
`z2 = np.arange(1,100)*100 # 100-m grid spacing`

## Get the grid spacing
`xs = np.around(np.unique(np.diff(x)).mean(), decimals=3)`
`ys = np.around(np.unique(np.diff(y)).mean(), decimals=3)`
`zs = np.around(np.unique(np.diff(z2)).mean(), decimals=3)`

`grid = pv.UniformGrid()`
`grid.spacing = xs, ys, zs`
`grid.origin = np.min(x), np.min(y), np.min(z2)`
`grid.dimensions = len(x), len(y), len(z)`
`print(grid.spacing)`
`print(grid.origin)`
`print(grid.dimensions)`
`grid`

(600.0, 600.0, 100.0)
(0.0, 0.0, 100.0)
(700, 600, 100)

<img width="334" alt="Screen Shot 2023-01-30 at 11 14 36 AM" src="https://user-images.githubusercontent.com/85080500/215531815-4c71bd70-8e7f-4daa-afd6-ccff13eeb156.png">

# Add 3d scalar field to grid
grid['twc'] = twc.ravel()
grid

<img width="752" alt="Screen Shot 2023-01-30 at 11 15 54 AM" src="https://user-images.githubusercontent.com/85080500/215532117-c60c9ece-5d30-4a7d-abc1-e44a9585b8dd.png">

# All good so far! But...

# Plot
`pv.set_jupyter_backend('static')`
`p = pv.Plotter(notebook=True, window_size=[1920, 1080])`

`p.add_volume(grid, scalars='twc',clim=[-6,1], cmap="bone_r", opacity="linear")`

`p.camera_position = 'xz'`
`p.camera.elevation = 20`
`p.camera.azimuth = 30`
`p.camera.zoom(1.5)`
`p.show_bounds(xlabel='x [m]',ylabel='y [m]',zlabel='Height [m]',location='outer',grid='front',font_family='arial')`
`p.show()`

# Produces the following, which looks like it's contouring something related to the grid
![download](https://user-images.githubusercontent.com/85080500/215529234-8f6a5d30-1816-4a01-b8dd-61d09398dc73.png)

# Now print the min and max of TWC and grid
`print(np.nanmax(twc))`
`print(np.nanmin(twc))`
`grid`

255.0
0.0
<img width="733" alt="Screen Shot 2023-01-30 at 11 19 37 AM" src="https://user-images.githubusercontent.com/85080500/215533159-3791c0a8-d6f3-4d72-97f8-ed6dc0c9351e.png">

# Now, the 'twc' scalar field has been replaced by the np.unit8 0-255 color range field. But if I use that anyways to plot as follows:

`pv.set_jupyter_backend('static')`
`p = pv.Plotter(notebook=True,window_size=[1920, 1080])`

`vol = twc.copy()`
`scalars=twc.copy()`
`#p.add_volume(grid, scalars='twc',clim=[-6,1], cmap="bone", opacity="linear")`
`p.add_volume(vol,scalars=scalars, clim=[-6,1], cmap="bone_r", opacity="linear")`

`p.camera_position = 'xz'`
`p.camera.elevation = 20`
`p.camera.azimuth = 30`
`p.camera.zoom(1.5)`

`p.show_bounds(xlabel='x [m]',ylabel='y [m]',zlabel='Height [m]',location='outer',grid='front',font_family='arial')`

`p.show()`

# I get the following, which is what I'm trying to achieve

![download-1](https://user-images.githubusercontent.com/85080500/215529278-3ccf53f3-c464-4490-bd08-390b2dbd547c.png)

# But clearly this isn't the correct way to do this, and it's not on the native grid anymore. Any pointers?
#----------------------------------------------------------

**Comment by banesullivan** [comment, marked as answer]
Can you share the source data file? The first plot you show looks like a standard C vs F ordering issue when raveling the array. I.e. try `grid['twc'] = twc.ravel(order='F')`

**Comment by mckenna-stanford-columbia** [comment]
The file size (netCDF) is too large to upload to my repository as I don't have git-lfs installed. But you were spot-on. This worked, as follows:
![download-3](https://user-images.githubusercontent.com/85080500/215545515-ad2fe72f-d201-4493-ba96-87e433fed7e5.png)

Thanks so much, I never would have thought of that.
